// @recharts/devtools/dist/annotations/AnnotationsController.js
/**
 * @fileoverview renders UI form for managing annotations in a chart.
 *
 * This is basic UI for demonstration purposes.
 * In your own application, this UI is probably going to look completely different.
 * Feel free to reuse what you see here but chances are you'll want to build your own anyway.
 *
 * Allows:
 *
 * - Adding, editing, and deleting annotations.
 * - Supported annotations: Crosshairs, Labels, Shapes (lines, rectangles, circles).
 * - Configuring annotation styles: color, strokeWidth, strokeDasharray, fill, opacity.
 * - Positioning annotations using coordinates. Allows both absolute pixel values and data values.
 * - Adding labels to annotations with customizable text and styles and position.
 *
 * This controller is meant to be rendered outside the chart area, typically in a sidebar or modal,
 * in an HTML context (not SVG).
 */
import React, { useId, useState } from 'react';
const ANNOTATION_TYPES = [
    { type: 'horizontalLine', label: 'Horizontal Line' },
    { type: 'verticalLine', label: 'Vertical Line' },
    { type: 'circle', label: 'Circle' },
    { type: 'rectangle', label: 'Rectangle' },
    { type: 'label', label: 'Label' },
    { type: 'freeformLine', label: 'Freeform Line' },
    { type: 'crosshair', label: 'Crosshair' },
];
const buttonStyle = {
    padding: '4px 8px',
    fontSize: '11px',
    cursor: 'pointer',
    border: '1px solid #ccc',
    borderRadius: '3px',
    background: '#fff',
};
const activeButtonStyle = {
    ...buttonStyle,
    background: '#e0e0e0',
    borderColor: '#999',
};
const inputStyle = {
    width: '50px',
    fontSize: '11px',
    padding: '2px 4px',
    border: '1px solid #ccc',
    borderRadius: '2px',
};
const labelStyle = {
    fontSize: '11px',
    minWidth: '40px',
    display: 'inline-flex',
    flexDirection: 'row',
    alignItems: 'center',
    gap: '1ex',
    marginRight: '2ex',
};
function SnapModeRadioGroup({ snapMode, onSnapModeChange, }) {
    const name = useId();
    return (React.createElement("div", { style: { marginBottom: '12px' } },
        React.createElement("div", { style: { marginBottom: '6px', fontWeight: 'bold' } }, "Snap to Data:"),
        React.createElement("label", { style: labelStyle },
            React.createElement("input", { type: "radio", name: name, value: "none", checked: snapMode === 'none', onChange: () => onSnapModeChange('none') }),
            "No snapping"),
        React.createElement("label", { style: labelStyle },
            React.createElement("input", { type: "radio", name: name, value: "data", checked: snapMode === 'data', onChange: () => onSnapModeChange('data') }),
            "Snap to data points"),
        React.createElement("label", { style: labelStyle },
            React.createElement("input", { type: "radio", name: name, value: "tick", checked: snapMode === 'tick', onChange: () => onSnapModeChange('tick') }),
            "Snap to axis ticks")));
}
function AnnotationEditForm({ annotation, handleStyleChange, onUpdateAnnotation, }) {
    return (React.createElement("div", { style: { marginTop: '8px', paddingTop: '8px', borderTop: '1px solid #ddd' } },
        React.createElement("div", { style: { display: 'flex', flexWrap: 'wrap', gap: '8px', marginBottom: '8px' } },
            (annotation.type === 'verticalLine' ||
                annotation.type === 'circle' ||
                annotation.type === 'label' ||
                annotation.type === 'crosshair') && (React.createElement("label", { style: labelStyle },
                "X:",
                React.createElement("input", { type: "number", value: annotation.pointA.interactionCoordinate.x, onChange: (e) => onUpdateAnnotation({
                        ...annotation,
                        pointA: {
                            ...annotation.pointA,
                            interactionCoordinate: {
                                ...annotation.pointA.interactionCoordinate,
                                x: parseFloat(e.target.value),
                            },
                        },
                    }), style: inputStyle }))),
            (annotation.type === 'horizontalLine' ||
                annotation.type === 'circle' ||
                annotation.type === 'label' ||
                annotation.type === 'crosshair') && (React.createElement("label", { style: labelStyle },
                "Y:",
                React.createElement("input", { type: "number", value: annotation.pointA.interactionCoordinate.y, onChange: (e) => onUpdateAnnotation({
                        ...annotation,
                        pointA: {
                            ...annotation.pointA,
                            interactionCoordinate: {
                                ...annotation.pointA.interactionCoordinate,
                                y: parseFloat(e.target.value),
                            },
                        },
                    }), style: inputStyle }))),
            (annotation.type === 'rectangle' || annotation.type === 'freeformLine') && (React.createElement(React.Fragment, null,
                React.createElement("label", { style: labelStyle },
                    "X1:",
                    React.createElement("input", { type: "number", value: annotation.pointA.interactionCoordinate.x, onChange: (e) => onUpdateAnnotation({
                            ...annotation,
                            pointA: {
                                ...annotation.pointA,
                                interactionCoordinate: {
                                    x: parseFloat(e.target.value),
                                    y: annotation.pointA.interactionCoordinate.y,
                                },
                            },
                        }), style: inputStyle })),
                React.createElement("label", { style: labelStyle },
                    "Y1:",
                    React.createElement("input", { type: "number", value: annotation.pointA.interactionCoordinate.y, onChange: (e) => onUpdateAnnotation({
                            ...annotation,
                            pointA: {
                                ...annotation.pointA,
                                interactionCoordinate: {
                                    x: annotation.pointA.interactionCoordinate.x,
                                    y: parseFloat(e.target.value),
                                },
                            },
                        }), style: inputStyle })),
                React.createElement("label", { style: labelStyle },
                    "X2:",
                    React.createElement("input", { type: "number", value: annotation.pointB?.interactionCoordinate.x ?? '', onChange: (e) => onUpdateAnnotation({
                            ...annotation,
                            pointB: {
                                snappedCoordinate: undefined,
                                dataPoint: undefined,
                                interactionCoordinate: {
                                    x: parseFloat(e.target.value),
                                    y: annotation.pointB?.interactionCoordinate.y ?? 0,
                                },
                            },
                        }), style: inputStyle })),
                React.createElement("label", { style: labelStyle },
                    "Y2:",
                    React.createElement("input", { type: "number", value: annotation.pointB?.interactionCoordinate.y ?? '', onChange: (e) => onUpdateAnnotation({
                            ...annotation,
                            pointB: {
                                snappedCoordinate: undefined,
                                dataPoint: undefined,
                                interactionCoordinate: {
                                    x: annotation.pointB?.interactionCoordinate.x ?? 0,
                                    y: parseFloat(e.target.value),
                                },
                            },
                        }), style: inputStyle }))))),
        React.createElement("div", { style: { display: 'flex', flexWrap: 'wrap', gap: '8px' } },
            React.createElement("label", { style: labelStyle },
                "Stroke:",
                React.createElement("input", { type: "number", value: annotation.style.strokeWidth ?? 2, onChange: (e) => handleStyleChange(annotation, 'strokeWidth', parseFloat(e.target.value)), style: inputStyle, min: 0, step: 0.5 })),
            React.createElement("label", { style: labelStyle },
                "Opacity:",
                React.createElement("input", { type: "number", value: annotation.style.opacity ?? 1, onChange: (e) => handleStyleChange(annotation, 'opacity', parseFloat(e.target.value)), style: inputStyle, min: 0, max: 1, step: 0.1 })),
            React.createElement("label", { style: labelStyle },
                "Dash:",
                React.createElement("input", { type: "text", value: annotation.style.strokeDasharray ?? '', onChange: (e) => handleStyleChange(annotation, 'strokeDasharray', e.target.value), style: { ...inputStyle, width: '60px' }, placeholder: "e.g. 4 2" })),
            (annotation.type === 'rectangle' || annotation.type === 'circle') && (React.createElement(React.Fragment, null,
                React.createElement("label", { style: labelStyle },
                    "Fill:",
                    React.createElement("input", { type: "color", value: annotation.style.fill ?? '#cccccc', onChange: (e) => handleStyleChange(annotation, 'fill', e.target.value), style: {
                            width: '20px',
                            height: '20px',
                            padding: 0,
                            border: 'none',
                            cursor: 'pointer',
                        } })),
                React.createElement("label", { style: labelStyle },
                    "Fill Opacity:",
                    React.createElement("input", { type: "number", value: annotation.style.fillOpacity ?? 0.1, onChange: (e) => handleStyleChange(annotation, 'fillOpacity', parseFloat(e.target.value)), style: inputStyle, min: 0, max: 1, step: 0.1 }))))),
        annotation.type === 'label' && (React.createElement("div", { style: { marginTop: '8px' } },
            React.createElement("label", { style: labelStyle },
                "Text:",
                React.createElement("input", { type: "text", value: String(annotation.label?.value ?? ''), onChange: (e) => onUpdateAnnotation({
                        ...annotation,
                        label: { ...annotation.label, value: e.target.value },
                    }), style: { ...inputStyle, width: '150px' } }))))));
}
/**
 * Renders controls for adding, editing, and removing annotations.
 *
 * This component renders HTML elements and is meant to be rendered via portal
 * outside the SVG chart context.
 */
export function AnnotationsController({ annotations, isAdding, selectedAnnotationId, snapMode, onStartAdding, onCancelAdding, onDeleteAnnotation, onUpdateAnnotation, onSelectAnnotation, onSnapModeChange, }) {
    const [expandedAnnotationId, setExpandedAnnotationId] = useState(null);
    const handleStyleChange = (annotation, styleKey, value) => {
        onUpdateAnnotation({
            ...annotation,
            style: {
                ...annotation.style,
                [styleKey]: value,
            },
        });
    };
    const toggleExpanded = (id) => {
        setExpandedAnnotationId((prev) => (prev === id ? null : id));
    };
    return (React.createElement("div", { style: { fontFamily: 'monospace', fontSize: '12px' } },
        React.createElement("div", { style: { marginBottom: '12px' } },
            React.createElement("div", { style: { marginBottom: '6px', fontWeight: 'bold' } }, "Add Annotation:"),
            React.createElement("div", { style: { display: 'flex', flexWrap: 'wrap', gap: '4px' } }, ANNOTATION_TYPES.map(({ type, label }) => (React.createElement("button", { key: type, type: "button", onClick: () => (isAdding === type ? onCancelAdding() : onStartAdding(type)), style: isAdding === type ? activeButtonStyle : buttonStyle }, isAdding === type ? `Cancel ${label}` : label)))),
            isAdding && (React.createElement("div", { style: { marginTop: '6px', fontSize: '11px', color: '#666' } },
                "Click on the chart to place the annotation",
                (isAdding === 'rectangle' || isAdding === 'freeformLine' || isAdding === 'circle') &&
                    ' (click twice or drag to define dimensions)'))),
        React.createElement(SnapModeRadioGroup, { snapMode: snapMode, onSnapModeChange: onSnapModeChange }),
        React.createElement("div", null,
            React.createElement("div", { style: { marginBottom: '6px', fontWeight: 'bold' } },
                "Annotations (",
                annotations.length,
                "):"),
            annotations.length === 0 ? (React.createElement("div", { style: { fontSize: '11px', color: '#666' } }, "No annotations yet")) : (React.createElement("div", { style: { display: 'flex', flexDirection: 'column', gap: '8px' } }, annotations.map((annotation) => (React.createElement("div", { key: annotation.id, style: {
                    padding: '8px',
                    border: selectedAnnotationId === annotation.id ? '2px solid #007bff' : '1px solid #ddd',
                    borderRadius: '4px',
                    background: '#fafafa',
                } },
                React.createElement("div", { style: { display: 'flex', alignItems: 'center', gap: '8px', marginBottom: '4px' } },
                    React.createElement("input", { type: "color", value: annotation.style.color ?? '#ff0000', onChange: (e) => handleStyleChange(annotation, 'color', e.target.value), style: {
                            width: '20px',
                            height: '20px',
                            padding: 0,
                            border: 'none',
                            cursor: 'pointer',
                        }, title: "Change color" }),
                    React.createElement("span", { style: { fontWeight: 'bold', fontSize: '11px', flex: 1 } }, ANNOTATION_TYPES.find((t) => t.type === annotation.type)?.label ??
                        annotation.type),
                    React.createElement("button", { type: "button", onClick: () => toggleExpanded(annotation.id), style: buttonStyle }, expandedAnnotationId === annotation.id ? '▼' : '▶'),
                    React.createElement("button", { type: "button", onClick: () => onSelectAnnotation(selectedAnnotationId === annotation.id ? null : annotation.id), style: selectedAnnotationId === annotation.id ? activeButtonStyle : buttonStyle }, selectedAnnotationId === annotation.id ? 'Deselect' : 'Select'),
                    React.createElement("button", { type: "button", onClick: () => onDeleteAnnotation(annotation.id), style: { ...buttonStyle, color: '#d00' } }, "Delete")),
                expandedAnnotationId === annotation.id && (React.createElement(AnnotationEditForm, { annotation: annotation, handleStyleChange: handleStyleChange, onUpdateAnnotation: onUpdateAnnotation }))))))))));
}
