// @recharts/devtools/dist/annotations/useAnnotationsManager.js
import { useCallback, useState } from 'react';
import { DEFAULT_ANNOTATION_STYLE, getAnnotationColor } from './annotationColors.js';
import { getDistance } from './getDistance.js';
import { DEFAULT_CIRCLE_RADIUS, DEFAULT_RECT_HEIGHT, DEFAULT_RECT_WIDTH, } from './annotationInteractionUtils.js';
function annotationFactory(type, color, snapMode, pointA, pointB) {
    const baseStyle = {
        ...DEFAULT_ANNOTATION_STYLE,
        color,
        fill: color,
        fillOpacity: 0.1,
    };
    return {
        type,
        id: Date.now(),
        positionType: snapMode === 'none' ? 'pixel' : 'data',
        style: baseStyle,
        pointA,
        pointB,
        label: type === 'label' ? { value: 'Label' } : undefined,
    };
}
/**
 * Hook for managing annotation state and interactions.
 *
 * This hook provides state management for annotations including:
 * - Adding, editing, and deleting annotations
 * - Tracking mouse position for annotation placement
 * - Selection state for editing
 */
export const useAnnotationsManager = (props) => {
    const [annotations, setAnnotations] = useState([]);
    const [isAdding, setIsAdding] = useState(null);
    const [followerPosition, setFollowerPosition] = useState(null);
    const [selectedAnnotationId, setSelectedAnnotationId] = useState(null);
    const [snapMode, setSnapMode] = useState('none');
    // Track click positions for multi-click annotations (rectangle, freeform line)
    const [firstClickPosition, setFirstClickPosition] = useState(null);
    const startAddingAnnotation = useCallback((type) => {
        setIsAdding(type);
        setFirstClickPosition(null);
        setFollowerPosition(null);
    }, []);
    const cancelAddingAnnotation = useCallback(() => {
        setIsAdding(null);
        setFirstClickPosition(null);
        setFollowerPosition(null);
    }, []);
    const deleteAnnotation = useCallback((id) => {
        setAnnotations((prev) => prev.filter((a) => a.id !== id));
        props?.onAnnotationDelete?.(id);
        if (selectedAnnotationId === id) {
            setSelectedAnnotationId(null);
        }
    }, [props, selectedAnnotationId]);
    const updateAnnotation = useCallback((updatedAnnotation) => {
        setAnnotations((prev) => prev.map((a) => (a.id === updatedAnnotation.id ? updatedAnnotation : a)));
        props?.onAnnotationUpdate?.(updatedAnnotation);
    }, [props]);
    const selectAnnotation = useCallback((id) => {
        setSelectedAnnotationId(id);
    }, []);
    const createAnnotation = useCallback((type, snapMode, pointA, pointB) => {
        const color = getAnnotationColor(annotations.length);
        return annotationFactory(type, color, snapMode, pointA, pointB);
    }, [annotations.length]);
    const onChartMouseDown = useCallback((snap) => {
        if (!isAdding)
            return;
        // For rectangle and freeform line, MouseDown sets the first point (anchor)
        if (isAdding === 'rectangle' || isAdding === 'freeformLine' || isAdding === 'circle') {
            if (!firstClickPosition) {
                setFirstClickPosition(snap);
            }
        }
    }, [isAdding, firstClickPosition]);
    const onChartMouseUp = useCallback((snap) => {
        if (!isAdding)
            return;
        // For rectangle and freeform line, MouseUp can complete the annotation
        // if it's a drag (distance > threshold) OR if it's the second click of a click-click flow
        if (isAdding === 'rectangle' || isAdding === 'freeformLine' || isAdding === 'circle') {
            if (firstClickPosition) {
                const dist = getDistance(firstClickPosition, snap);
                let expectedMinimumDistance = 3;
                // In pixel mode the anchor is offset from the real cursor by the default annotation
                // size, so we raise the threshold to avoid confirming on the first stationary click.
                // In data/tick mode there is no offset, so the base 3px threshold is sufficient.
                if (snapMode === 'none') {
                    switch (isAdding) {
                        case 'circle': {
                            expectedMinimumDistance += DEFAULT_CIRCLE_RADIUS;
                            break;
                        }
                        case 'rectangle': {
                            expectedMinimumDistance += getDistance(snap, {
                                interactionCoordinate: {
                                    x: snap.interactionCoordinate.x + DEFAULT_RECT_WIDTH,
                                    y: snap.interactionCoordinate.y + DEFAULT_RECT_HEIGHT,
                                },
                                snappedCoordinate: undefined,
                                dataPoint: undefined,
                            });
                            break;
                        }
                    }
                }
                // Threshold to distinguish between a simple click (to set anchor) and a drag/second click
                if (dist > expectedMinimumDistance) {
                    const newAnnotation = createAnnotation(isAdding, snapMode, firstClickPosition, snap);
                    setAnnotations((prev) => [...prev, newAnnotation]);
                    props?.onAnnotationAdd?.(newAnnotation);
                    setIsAdding(null);
                    setFirstClickPosition(null);
                    setFollowerPosition(null);
                }
            }
        }
    }, [isAdding, firstClickPosition, createAnnotation, props, snapMode]);
    const onChartClick = useCallback((snap) => {
        if (!isAdding)
            return;
        // Two-point annotations are now handled by MouseDown/MouseUp
        if (isAdding === 'rectangle' || isAdding === 'freeformLine' || isAdding === 'circle') {
            return;
        }
        // For other annotation types, single click creates the annotation
        const newAnnotation = createAnnotation(isAdding, snapMode, snap);
        setAnnotations((prev) => [...prev, newAnnotation]);
        props?.onAnnotationAdd?.(newAnnotation);
        setIsAdding(null);
        setFollowerPosition(null);
    }, [isAdding, createAnnotation, props, snapMode]);
    const onChartMouseMove = useCallback((snap) => {
        if (isAdding) {
            setFollowerPosition(snap);
        }
    }, [isAdding]);
    const onChartMouseLeave = useCallback(() => {
        setFollowerPosition(null);
    }, []);
    return {
        annotations,
        isAdding,
        followerPosition,
        firstClickPosition,
        selectedAnnotationId,
        snapMode,
        startAddingAnnotation,
        cancelAddingAnnotation,
        deleteAnnotation,
        updateAnnotation,
        selectAnnotation,
        setSnapMode,
        onChartClick,
        onChartMouseDown,
        onChartMouseUp,
        onChartMouseMove,
        onChartMouseLeave,
    };
};
